scene/resources/svg_texture.cpp

```diff
@@ -0,0 +1,381 @@
+	return rid;
+}
+
+void SVGTexture::_clear() {
+	for (KeyValue<double, RID> &tx : texture_cache) {
+		if (tx.value.is_valid()) {
+			RenderingServer::get_singleton()->free(tx.value);
+		}
+	}
+	texture_cache.clear();
+	if (base_texture.is_valid()) {
+		RenderingServer::get_singleton()->free(base_texture);
+	}
+	base_texture = RID();
+	alpha_cache.unref();
+}
+
+void SVGTexture::_update_texture() {
+	_clear();
+	_ensure_scale(1.0);
+
+	emit_changed();
+}
+
+Ref<Image> SVGTexture::get_image() const {
+	RID rid = _ensure_scale(1.0);
+	if (rid.is_valid()) {
+		return RenderingServer::get_singleton()->texture_2d_get(rid);
+	} else {
+		return Ref<Image>();
+	}
+}
+
+int SVGTexture::get_width() const {
+	return size.x;
+}
+
+int SVGTexture::get_height() const {
+	return size.y;
+}
+
+RID SVGTexture::get_rid() const {
+	return _ensure_scale(1.0);
+}
+
+bool SVGTexture::has_alpha() const {
+	return true;
+}
+
+void SVGTexture::draw(RID p_canvas_item, const Point2 &p_pos, const Color &p_modulate, bool p_transpose) const {
+	double scale = 1.0;
+	CanvasItem *ci = CanvasItem::get_current_item_drawn();
+	if (ci) {
+		Viewport *vp = ci->get_viewport();
+		if (vp) {
+			scale = vp->get_oversampling();
+		}
+	}
+	RID rid = _ensure_scale(scale);
+
+	RenderingServer::get_singleton()->canvas_item_add_texture_rect(p_canvas_item, Rect2(p_pos, size), rid, false, p_modulate, p_transpose);
+}
+
+void SVGTexture::draw_rect(RID p_canvas_item, const Rect2 &p_rect, bool p_tile, const Color &p_modulate, bool p_transpose) const {
+	double scale = 1.0;
+	CanvasItem *ci = CanvasItem::get_current_item_drawn();
+	if (ci) {
+		Viewport *vp = ci->get_viewport();
+		if (vp) {
+			scale = vp->get_oversampling();
+		}
+	}
+	RID rid = _ensure_scale(scale);
+
+	RenderingServer::get_singleton()->canvas_item_add_texture_rect(p_canvas_item, p_rect, rid, p_tile, p_modulate, p_transpose);
+}
+
+void SVGTexture::draw_rect_region(RID p_canvas_item, const Rect2 &p_rect, const Rect2 &p_src_rect, const Color &p_modulate, bool p_transpose, bool p_clip_uv) const {
+	double scale = 1.0;
+	CanvasItem *ci = CanvasItem::get_current_item_drawn();
+	if (ci) {
+		Viewport *vp = ci->get_viewport();
+		if (vp) {
+			scale = vp->get_oversampling();
+		}
+	}
+	RID rid = _ensure_scale(scale);
+
+	RenderingServer::get_singleton()->canvas_item_add_texture_rect_region(p_canvas_item, p_rect, rid, p_src_rect, p_modulate, p_transpose, p_clip_uv);
+}
+
+bool SVGTexture::is_pixel_opaque(int p_x, int p_y) const {
+	if (alpha_cache.is_null()) {
+		Ref<Image> img = get_image();
+		if (img.is_valid()) {
+			alpha_cache.instantiate();
+			alpha_cache->create_from_image_alpha(img);
+		}
+	}
+
+	if (alpha_cache.is_valid()) {
+		int aw = int(alpha_cache->get_size().width);
+		int ah = int(alpha_cache->get_size().height);
+		if (aw == 0 || ah == 0) {
+			return true;
+		}
+
+		int x = p_x * aw / size.x;
+		int y = p_y * ah / size.y;
+
+		x = CLAMP(x, 0, aw);
+		y = CLAMP(y, 0, ah);
```

Comment:
Shouldn't it be aw - 1, ah - 1?

Reply:
Looks like it, unless `BitMap` have some special index logic, I'll check it later. This code is copy of `ImageTexture` (and `CompressedTexture2D`/`PortableCompressedTexture2D`) code:

https://github.com/godotengine/godot/blob/2d3bdcac35ac22ee6c4d5d5edc88ba947d0659d5/scene/resources/image_texture.cpp#L190-L204

Reply:
https://github.com/godotengine/godot/pull/105561